### PR TITLE
Contracts support

### PR DESCRIPTION
**Pull request by mestachs**
So the idea is to have contracts as simple events. 

To fetch them efficiently, we assume an SQLView has been setuped with something like:

To ease the migration path and ensure we can turn the contratct into "groups" for "lastest" period.
I assume some naming convention between optionSet/dataElements and orgUnitGroups/orgUnitGroupSets

image

image

Note that 
- the contract_start_date and contract_start_end should be there, 
- the contract_types are defined by the pbf scheme (number of dataElements, names, values are "open"). 
- If it's a contract/subonctract the contract_main_orgunit should be the name of the dataElement.

For the sqlView and client code, the dhis2 version is currently assumed as 2.32 or newer.
And it's configured without cache as a "simple" query : 
```sql
select 
programstageinstance.uid as event_id, 
programstageinstance.executiondate as event_date, 
program.uid as program_id, 
programStage.uid as program_stage_id,
programstageinstance.eventdatavalues as data_values, 
organisationunit.uid  as org_unit_id, 
organisationunit.name as org_unit_name, 
organisationunit.path as org_unit_path,
_orgunitstructure.*
from programstageinstance 
join programstage ON programstage.programstageid = programstageinstance.programstageid join program ON program.programid = programstage.programid 
join organisationunit ON organisationunit.organisationunitid = programstageinstance.organisationunitid 
join _orgunitstructure ON _orgunitstructure.organisationunitid = programstageinstance.organisationunitid 
where  program.uid = '${programId}' and programstageinstance.deleted is false order by programstageinstanceid desc 
```

## Self proof reading checklist

- [ ] Did I run Pronto and fixed all warnings
- [ ] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [ ] Did I test the right thing?
- [ ] Did I test corner cases, non happy path (defensive testing)?
- [ ] Check that I used the ad-hoc patterns and created files accordingly
- [ ] Check code efficiency with record intensive project
- [ ] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

File: lib/orbf/rules_engine/data/contract.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        @field_values = field_values
+        @id = field_values["id"]
+        @org_unit = field_values["org_unit"]
+        @start_period = field_values["startContract"].gsub("-", "").slice(0, 6)
+        @end_period = field_values["endContract"].gsub("-", "").slice(0, 6)
+        @calendar = calendar
+      end
+
+      def match_period?(period)
+        start_month_period = @calendar.periods(period, "monthly")[0]
```
Comment (mestachs) [inline lib/orbf/rules_engine/data/contract.rb:16]: I think I'll need to change this to "some periods" and not only the first period

File: lib/orbf/rules_engine/fetch_data/resolve_arguments.rb [hunk 2]
```diff
@@ -26,22 +27,27 @@ def call
 
       def build_package_arguments
         project.packages.each_with_object({}) do |package, hash|
-          next unless match_group?(package)
+
+          orgunits = decorate_with_facts(build_orgunit_resolver(package, pyramid, main_orgunit).call)
```
Comment (mestachs) [inline lib/orbf/rules_engine/fetch_data/resolve_arguments.rb:31]: the decoration should probably use the contract info now

File: lib/orbf/rules_engine/services/contract_service.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      attr_reader :dhis2_connection
+
+      def to_event(row)
+        data_vals = JSON.parse(row[1]["value"])
+        data_values = data_vals.keys.map do |k|
+          data_vals[k]["dataElement"] = k
+          data_vals[k]
+        end
+        {
+          "event"        => row[0],
```
Comment (mestachs) [inline lib/orbf/rules_engine/services/contract_service.rb:82]: find a way to not hard code the indexes, see if there's some "usable" meta data in the answer.

File: spec/contracts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+require_relative "../lib/orbf/rules_engine"
+require "byebug"
+
+if ENV["DHIS2_PASSWORD"]
```
Comment (mestachs) [inline spec/contracts_spec.rb:4]: turn this in a real spec

File: lib/orbf/rules_engine/data/contract.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+      def org_unit_id
+        @org_unit["id"]
+      end
+
+      def org_unit_name
+        @org_unit["name"]
+      end
+
+      def codes
```
Comment (vanakenm) [inline lib/orbf/rules_engine/data/contract.rb:30]: That's to get all codes of groups (ie fields that are not the 3 "main ones") ?

Comment (mestachs) [reply lib/orbf/rules_engine/data/contract.rb:30]: yes but "open to more data elements/optionsSets", the lowercase is to be compliant with the groups returned by the pyramid.

File: lib/orbf/rules_engine/services/contract_service.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+      def mappings
+        @mappings ||= begin
+          data_elements = program.program_stages.flat_map do |ps|
+            ps["program_stage_data_elements"].map do |psde|
+              psde["data_element"]
+            end
+          end
+
+          data_elements.each_with_object({}) { |de, mappings| mappings[de["id"]] = de }
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/contract_service.rb:30]: Don't think it will ever be a (performance) problem, but you could do the flat_map and ~inject~ each_with_object in one `each_with_object` and save a loop

File: lib/orbf/rules_engine/services/contract_service.rb (hunk 3 above)
Comment (pjaspers) [reply lib/orbf/rules_engine/services/contract_service.rb:82]: (from the spec data it looks like the `headers` might have it)

File: lib/orbf/rules_engine/fetch_data/resolve_arguments.rb [hunk 7]
```diff
@@ -26,22 +27,27 @@ def call
...
         end
       end
 
-      def match_group?(package)
-        (package.main_org_unit_group_ext_ids & main_orgunit.group_ext_ids).any?
+      def build_orgunit_resolver(package, pyramid, main_orgunit)
+        if @contract_service
+          ContractOrgunitsResolver.new(package, pyramid, main_orgunit, @contract_service, invoicing_period)
+        else
+          OrgunitsResolver.new(package, pyramid, main_orgunit)
```
Comment (pjaspers) [inline lib/orbf/rules_engine/fetch_data/resolve_arguments.rb:49]: Thinking out loud here, would it make sence to move the `if @contract_service` inside the `OrgunitsResolver` and let that one use the `ContractOrgunitsResolver` when needed?

File: lib/orbf/rules_engine/fetch_data/contract_orgunits_resolver.rb [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        @package = package
+        @pyramid = pyramid
+        @main_orgunit = main_orgunit
+        @contract_service = contract_service
+        @main_orgunit_contract = contract_service.for(main_orgunit.ext_id, period)
+        @period = period
+      end
+
+      def call
+        unless @main_orgunit_contract
```
Comment (pjaspers) [inline lib/orbf/rules_engine/fetch_data/contract_orgunits_resolver.rb:16]: This is the 'no contract in place for this period' escape clause right? Should we worry/know about this, if we try to run it and we don't have a contract in place but we do have a contract_service defined on the project?

File: lib/orbf/rules_engine/fetch_data/resolve_arguments.rb (hunk 7 above)
Comment (mestachs) [reply lib/orbf/rules_engine/fetch_data/resolve_arguments.rb:49]: I wanted to keep things "seperated". and the existing OrgunitsResolver is allready complex.
and wanted to avoid passing "unused" params like invoicing_period and 

I may propose : 
- rename OrgunitsResolver to GroupsOrgunitsResolver 
- have a fresh OrgunitsResolver doing the switch between GroupsOrgunitsResolver and ContractOrgunitsResolver

File: lib/orbf/rules_engine/fetch_data/contract_orgunits_resolver.rb [hunk 9]
```diff
@@ -38,17 +43,14 @@ def handle_zone
...
         [main_orgunit]
       end
 
       def handle_target_org_units
-        target_codes = pyramid.groups(package.target_org_unit_group_ext_ids).map(&:code)
+        target_codes = pyramid.groups(package.target_org_unit_group_ext_ids).map(&:code_downcase)
         target_contracts = @contract_service.for_groups(target_codes, @period)
         org_units_set = pyramid.org_units_for(target_contracts.map(&:org_unit_id))
-        org_units_set = org_units_set.keep_if { |orgunit| orgunit.path.start_with?(main_orgunit.path) }
+        org_units_set = org_units_set.keep_if { |orgunit| orgunit.path ? orgunit.path.start_with?(main_orgunit.path) : byebug}
```
Comment (pjaspers) [inline lib/orbf/rules_engine/fetch_data/contract_orgunits_resolver.rb:53]: A lone `byebug` appeared

File: lib/orbf/rules_engine/services/groups_synchro.rb [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    class GroupsSynchro
+      attr_reader :contract_service
+
+      def initialize(contract_service)
+        @contract_service = contract_service
+        @groups = {}
+      end
+
+      def synchronize(period)
+        contract_by_orgunits = contract_service.find_all.group_by(&:org_unit_id)
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/groups_synchro.rb:14]: `contracts_by_orgunit` seems more correct

File: lib/orbf/rules_engine/services/groups_synchro.rb [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+      def dhis2
+        @dhis2 ||= contract_service.dhis2_connection
+      end
+
+      def find_group(code)
+        @groups[code] ||= begin
+          ou_groups = dhis2.organisation_unit_groups.list(filter: "code:eq:#{code}", fields: :all)
+          if ou_groups.empty?
+            raise "no organisation unit group with code #{code}"
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/groups_synchro.rb:80]: Not sure if it should raise?

If the group doesn't exist we can let it continue and the `remove_from` will just continue to work, otherwise we stop the complete sync.

File: lib/orbf/rules_engine/services/groups_synchro.rb [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        display_groups_stats("before")
+        @groups.each do |(_code, group)|
+          group.update
+        end
+        display_groups_stats("after")
+      end
+
+      def display_groups_stats(message)
+        stats = dhis2.get("organisationUnitGroupSets.json?fields=id,name,code,organisationUnitGroups[id,name,code,organisationUnits~size]")
+        puts ("*** Groups stats #{message} :" + JSON.pretty_generate(stats))
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/groups_synchro.rb:105]: I'm also using `JSON.pretty_generate` more and more to quickly show data!

File: lib/orbf/rules_engine/services/groups_synchro.rb [hunk 13]
```diff
@@ -0,0 +1,109 @@
+module Orbf
+  module RulesEngine
+    class BadDhis2ConfigError < StandardError
```
Comment (pjaspers) [inline lib/orbf/rules_engine/services/groups_synchro.rb:3]: Is this in use?

File: lib/orbf/rules_engine/services/groups_synchro.rb (hunk 11 above)
Comment (mestachs) [reply lib/orbf/rules_engine/services/groups_synchro.rb:80]: since it's a "toggle" between groups to another, it will break... I prefer the mapping between group codes and data elements codes to be 100% perfect and fail "fast"
